src/cdomains/lval.ml

```diff
@@ -628,3 +628,32 @@ struct
     end
     )
 end
+
+module OffsetNoIdx =
+struct
+  type someidx = SomeIdx [@@deriving eq, ord, hash]
+
+  include Printable.StdLeaf
+  type t = (CilType.Fieldinfo.t, someidx) offs [@@deriving eq, ord, hash]
```

Comment:
I'm really not a fan of yet another offset and value type. We already have way too many of them and all the copy-pasted conversions all over the place...

Address domain bucket representatives already are equivalent to this and avoid defining a custom `unit` type. Also access domain implements another equivalent copy. And there's probably more.

Although I still think it would be best for consistency to just use the same mutex lvalues type for may locks as we already have used for must locks for a decade.

Reply:
> Although I still think it would be best for consistency to just use the same mutex lvalues type for may locks as we already have used for must locks for a decade.

The problem is the offsets from `IdxDomain` that are used in these. I do not want to have to iterate over all buckets to see which indexes overlap and the join these.

Reply:
One could use the same type and replace all index offsets with unknown, but I am a bit wary of relying on such things if they are not expressed in the type as someone could easily change that back.

Reply:
> The problem is the offsets from `IdxDomain` that are used in these. I do not want to have to iterate over all buckets to see which indexes overlap and the join these.

Doesn't that mean the must locks analysis already has some issue with checking index aliasing? Because the improved version I mentioned above specifically keeps constant integer offsets around.

> One could use the same type and replace all index offsets with unknown, but I am a bit wary of relying on such things if they are not expressed in the type as someone could easily change that back.

The bucket representative offsets already implement a fully isomorphic type though with precisely the same meaning of a `unit` index for any index. There's no reason not to just reuse that module.

Reply:
I am not sure how much reuse we can get with that though, as it works for general addresses, and I do not want `NullPtr` or `StrPtr` to be possible values.

~~~OCaml
  (** Representatives for lvalue sublattices as defined by {!NormalLat}. *)
  module R: DisjointDomain.Representative with type elt = t =
  struct
    type elt = t

    (* Offset module for representative without abstract values for index offsets, i.e. with unit index offsets.
       Reason: The offset in the representative (used for buckets) should not depend on the integer domains,
       since different integer domains may be active at different program points. *)
    include Normal (UnitIdxDomain)

    let rec of_elt_offset: (fieldinfo, Idx.t) offs -> (fieldinfo, UnitIdxDomain.t) offs =
      function
      | `NoOffset -> `NoOffset
      | `Field (f,o) -> `Field (f, of_elt_offset o)
      | `Index (_,o) -> `Index (UnitIdxDomain.top (), of_elt_offset o) (* all indices to same bucket *)

    let of_elt (x: elt): t = match x with
      | Addr (v, o) -> Addr (v, of_elt_offset o) (* addrs grouped by var and part of offset *)
      | StrPtr _ when GobConfig.get_bool "ana.base.limit-string-addresses" -> StrPtr None (* all strings together if limited *)
      | StrPtr x -> StrPtr x (* everything else is kept separate, including strings if not limited *)
      | NullPtr -> NullPtr
      | UnknownPtr -> UnknownPtr
  end
~~~

Reply:
@sim642: Is that what you had in mind? https://github.com/goblint/analyzer/pull/839/commits/01b9841b262a911f79d30328d3d9b245408cc309